pkg/controller/kfservice/resources/knative/service.go

```diff
@@ -171,3 +170,81 @@ func (c *ServiceBuilder) CreatePredictorService(name string, metadata metav1.Obj
 
 	return service, nil
 }
+
+func (c *ServiceBuilder) CreateTransformerService(name string, metadata metav1.ObjectMeta, transformerSpec *v1alpha2.TransformerSpec, isCanary bool) (*knservingv1alpha1.Service, error) {
+	annotations := utils.Filter(metadata.Annotations, func(key string) bool {
+		return !utils.Includes(serviceAnnotationDisallowedList, key)
+	})
+
+	if transformerSpec.MinReplicas != 0 {
+		annotations[autoscaling.MinScaleAnnotationKey] = fmt.Sprint(transformerSpec.MinReplicas)
+	}
+	if transformerSpec.MaxReplicas != 0 {
+		annotations[autoscaling.MaxScaleAnnotationKey] = fmt.Sprint(transformerSpec.MaxReplicas)
+	}
+
+	// User can pass down scaling target annotation to overwrite the target default 1
+	if _, ok := annotations[autoscaling.TargetAnnotationKey]; !ok {
+		annotations[autoscaling.TargetAnnotationKey] = constants.DefaultScalingTarget
+	}
+	// User can pass down scaling class annotation to overwrite the default scaling KPA
+	if _, ok := annotations[autoscaling.ClassAnnotationKey]; !ok {
+		annotations[autoscaling.ClassAnnotationKey] = autoscaling.KPA
+	}
+
+	predict_url := constants.DefaultServiceName(metadata.Name, constants.Predictor) + "." + metadata.Namespace
+	if isCanary {
+		predict_url = constants.CanaryServiceName(metadata.Name, constants.Predictor) + "." + metadata.Namespace
+	}
+	service := &knservingv1alpha1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: metadata.Namespace,
+			Labels:    metadata.Labels,
+		},
+		Spec: knservingv1alpha1.ServiceSpec{
+			ConfigurationSpec: knservingv1alpha1.ConfigurationSpec{
+				Template: &knservingv1alpha1.RevisionTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: utils.Union(metadata.Labels, map[string]string{
+							constants.KFServicePodLabelKey: metadata.Name,
+						}),
+						Annotations: annotations,
+					},
+					Spec: knservingv1alpha1.RevisionSpec{
+						RevisionSpec: v1beta1.RevisionSpec{
+							// Defaulting here since this always shows a diff with nil vs 300s(knative default)
+							// we may need to expose this field in future
+							TimeoutSeconds: &constants.DefaultTimeout,
+							PodSpec: v1.PodSpec{
+								ServiceAccountName: transformerSpec.ServiceAccountName,
+								Containers: []v1.Container{
+									{
+										Image: transformerSpec.Custom.Container.Image,
+										Args: []string{
+											"--model_name",
+											metadata.Name,
+											"--predict_url",
+											"http://" + predict_url + "/v1/models/" + metadata.Name + ":predict",
```

Comment:
seems odd that the variable is called predict_url but its only hostname?

Reply:
still WIP :) was discussing with @cliveseldon to move this to a shared function returning the url.

Reply:
@rakelkar @cliveseldon I thought about this again, maybe it is not a good idea to construct the url here since url path is part of the data plane, we may not want the dependency with control plane to be aware of the data plane protocol. Model server should have the knowledge about the url path, so it can construct the full url there and control plane just pass the predictor hostname ?

Reply:
+1

Reply:
>  Model server should have the knowledge about the url path, so it can construct the full url there
What do you mean by the model server? This value is being injected into the transformer - so do you mean the transformer code?

Reply:
You could get the URL from the model framework classes (framework.go) so that its a detail of the predictor. The current implementation appears to have this encoded in the transformer. Would be nice to have the transformer be agnostic as anything in the py code gets baked into user artifacts and is difficult to change?